File: cpp/src/ble/ble_advertisement.cpp
#include "ble/ble_advertisement.h"

/// @brief Namespace for the layrz protocol
/// This namespace contains all the classes and methods related to the layrz
/// protocol
/// @copyright Copyright (c) 2025
namespace layrz_protocol
{
/// @brief Construct a new BleAdvertisement object
/// @param raw string containing the advertisement data
/// @return BleAdvertisement
BleAdvertisement::BleAdvertisement (
    const std::string &mac_address,
    const std::chrono::system_clock::time_point &timestamp,
    const std::optional<float> &latitude,
    const std::optional<float> &longitude,
    const std::optional<float> &altitude, const std::string &model,
    const std::optional<int> &tx_power,
    const std::vector<BleManufacturerData> &manufacturer_data,
    const std::vector<BleServiceData> &service_data, int rssi)
{
  this->mac_address = mac_address;
  this->timestamp = timestamp;
  this->latitude = latitude;
  this->longitude = longitude;
  this->altitude = altitude;
  this->model = model;
  this->tx_power = tx_power;
  this->manufacturer_data = manufacturer_data;
  this->service_data = service_data;
  this->rssi = rssi;
}
/// @brief Construct a new BleAdvertisement object
/// @param raw string containing the advertisement data
/// @return
BleAdvertisement::BleAdvertisement (const std::string &raw)
{
  *this = from_packet (raw);
}
/// @brief Construct a new BleAdvertisement object
/// @param raw string containing the advertisement data
/// @return BleAdvertisement
BleAdvertisement
BleAdvertisement::from_packet (const std::string &raw)
{
  std::vector<std::string> parts;
  std::stringstream ss (raw);
  std::string item;
  // Split the string by the semicolon
  while (std::getline (ss, item, ';'))
    parts.push_back (item);
  // Check if the packet has the correct number of fields
  if (parts.size () != 11)
    throw MalformedException ("Invalid packet definition");
  // Extract the fields
  auto raw_mac_address = parts[0];
  auto raw_timestamp = parts[1];
  auto raw_latitude = parts[2];
  auto raw_longitude = parts[3];
  auto raw_altitude = parts[4];
  auto raw_model = parts[5];
  auto raw_rssi = parts[6];
  auto raw_tx_power = parts[7];
  auto raw_manufacturer_data = parts[8];
  auto raw_service_data = parts[9];
  auto raw_crc = parts[10];
  // Convert the CRC to an integer
  int received_crc;
  try
    {
      received_crc = std::stoi (raw_crc, nullptr, 16);
    }
  catch (const std::invalid_argument &)
    {
      received_crc = 0;
    }
  // Check the CRC
  std::string sub_str = raw.substr (0, raw.size () - raw_crc.size () - 1);
  std::vector<uint8_t> bytes (sub_str.begin (), sub_str.end ());
  int calculated_crc = helpers::calculate_crc (bytes);
  // If the CRC is invalid, throw an exception
  if (received_crc != calculated_crc)
    throw CrcException ("Invalid CRC", received_crc, calculated_crc);
  // Check the MAC address
  if (raw_mac_address.size () != 12)
    throw MalformedException ("Invalid MAC Address");
  // Convert the fields to the correct types
  std::string mac_address;
  for (size_t i = 0; i < raw_mac_address.size (); i += 2)
    {
      if (i > 0)
        mac_address += ":";
      mac_address += raw_mac_address.substr (i, 2);
    }
  // Convert the timestamp to a time_point
  std::chrono::system_clock::time_point timestamp;
  try
    {
      timestamp = std::chrono::system_clock::from_time_t (
          std::stoll (raw_timestamp));
    }
  catch (const std::invalid_argument &)
    {
      throw MalformedException (
          "Invalid timestamp, should be an int or float");
    }
  // Convert the latitude to a float
  std::optional<float> latitude;
  if (!raw_latitude.empty ())
    {
      try
        {
          latitude = std::stof (raw_latitude);
        }
      catch (const std::invalid_argument &)
        {
          throw MalformedException ("Invalid latitude, should be a float");
        }
    }
  // Convert the longitude to a float
  std::optional<float> longitude;
  if (!raw_longitude.empty ())
    {
      try
        {
          longitude = std::stof (raw_longitude);
        }
      catch (const std::invalid_argument &)
        {
          throw MalformedException ("Invalid longitude, should be a float");
        }
    }
  // Convert the altitude to a float
  std::optional<float> altitude;
  if (!raw_altitude.empty ())
    {
      try
        {
          altitude = std::stof (raw_altitude);
        }
      catch (const std::invalid_argument &)
        {
          throw MalformedException ("Invalid altitude, should be a float");
        }
    }
  // Convert the RSSI to an int
  int rssi;
  try
    {
      rssi = std::stoi (raw_rssi);
    }
  catch (const std::invalid_argument &)
    {
      throw MalformedException ("Invalid RSSI, should be an int");
    }
  // Convert the TX Power to an int
  std::optional<int> tx_power;
  if (!raw_tx_power.empty ())
    {
      try
        {
          tx_power = std::stoi (raw_tx_power);
        }
      catch (const std::invalid_argument &)
        {
          throw MalformedException ("Invalid TX Power, should be an int");
        }
    }
  // Convert the manufacturer data to a vector of BleManufacturerData
  std::vector<BleManufacturerData> manufacturer_data;
  std::stringstream ss_manufacturer (raw_manufacturer_data);
  while (std::getline (ss_manufacturer, item, ','))
    {
      BleManufacturerData ble = BleManufacturerData (item);
      // ble.from_packet (item);
      manufacturer_data.push_back (ble);
    }

  std::vector<BleServiceData> service_data;
  std::stringstream ss_service (raw_service_data);
  while (std::getline (ss_service, item, ','))
    {
      service_data.push_back (BleServiceData::from_packet (item));
    }

  BleAdvertisement ble = BleAdvertisement (
      mac_address, timestamp, latitude, longitude, altitude, raw_model,
      tx_power, manufacturer_data, service_data, rssi);
  return ble;
}

/// @brief Convert the BleAdvertisement object to a packet string
/// @return std::string The packet string
std::string
BleAdvertisement::to_packet () const noexcept
{
  std::stringstream raw;
  raw << mac_address;
  raw << ";"
      << std::chrono::duration_cast<std::chrono::seconds> (
             timestamp.time_since_epoch ())
             .count ();
  raw << ";" << (latitude ? std::to_string (*latitude) : "");
  raw << ";" << (longitude ? std::to_string (*longitude) : "");
  raw << ";" << (altitude ? std::to_string (*altitude) : "");
  raw << ";" << model;
  raw << ";" << rssi;
  raw << ";" << (tx_power ? std::to_string (*tx_power) : "");
  raw << ";" << helpers::str::join (manufacturer_data, ",");
  raw << ";" << helpers::str::join (service_data, ",");
  raw << ";";

  std::vector<uint8_t> raw_bytes (raw.str ().begin (), raw.str ().end ());
  int crc = helpers::calculate_crc (raw_bytes);
  raw << std::hex << std::uppercase << std::setw (4) << std::setfill ('0')
      << crc;

  return raw.str ();
}
} // namespace layrz_protocol

File: dart/lib/src/packets/src/utils/ble_advertisement.dart
part of '../../packets.dart';

class BleData {
  /// [macAddress] is the detected Mac Address. This Mac Adress came from the detected device, not the
  /// device that is detecting.
  final String macAddress;

  /// [model] is the model of the detected device. This model should be equals to the model of the device
  /// and the model defined by Layrz.
  final String model;

  /// [BleData] is the BLE data packet.
  ///
  /// This packet is part of the packet sent from the server to the device.
  const BleData({
    required this.macAddress,
    required this.model,
  });

  /// [fromPackets] creates a [List<BleData>] from a raw message following this structure:
  /// MAC_ADDRESS:MODEL
  static List<BleData> fromPackets(String raw) {
    final parts = raw.split(';');
    if (parts.isEmpty) return [];

    if (parts.length == 1) return [];
    // The parts should be divisible by 2, can be multiple groups

    // Separate each group
    final List<BleData> devices = [];

    for (int i = 0; i < parts.length; i++) {
      final part = parts[i].split(':');
      if (part.length != 2) {
        throw MalformedException('Invalid BLE data definition');
      }

      final rawMacAddress = part[0];
      String macAddress = '';
      for (int i = 0; i < rawMacAddress.length; i += 2) {
        macAddress += '${rawMacAddress[i]}${rawMacAddress[i + 1]}';
        if (i + 2 < rawMacAddress.length) {
          macAddress += ':';
        }
      }

      final model = part[1];

      devices.add(BleData(macAddress: macAddress, model: model));
    }

    return devices;
  }

  /// [toPacket] converts the [BleData] to a raw message following this structure:
  /// MAC_ADDRESS:MODEL
  String toPacket() {
    return '${macAddress.replaceAll(':', '').toUpperCase()}:$model';
  }
}

class BleAdvertisement {
  /// [macAddress] is the detected Mac Address. This Mac Adress came from the detected device, not the
  /// device that is detecting.
  final String macAddress;

  /// [timestamp] is when the device was detected.
  final DateTime timestamp;

  /// [latitude] is the closest latitude of the device. Defined by the device that is detecting.
  /// This value is optional
  final double? latitude;

  /// [longitude] is the closest longitude of the device. Defined by the device that is detecting.
  /// This value is optional
  final double? longitude;

  /// [altitude] is the closest altitude of the device. Defined by the device that is detecting.
  /// This value is optional
  final double? altitude;

  /// [rssi] is the signal strength of the detected device.
  final int rssi;

  /// [txPower] is the transmission power of the detected device.
  /// This value is optional
  final int? txPower;

  /// [model] is the model of the detected device. This model should be equals to the model of the device
  /// and the model defined by Layrz.
  final String model;

  /// [manufacturerData] is the list of manufacturer data advertised by the device.
  List<BleManufacturerData> manufacturerData;

  /// [serviceData] is the list of service data advertised by the device.
  List<BleServiceData> serviceData;

  /// [deviceName] is the name of the device. This name is optional.
  final String? deviceName;

  BleAdvertisement({
    required this.macAddress,
    required this.timestamp,
    this.latitude,
    this.longitude,
    this.altitude,
    required this.rssi,
    this.txPower,
    required this.model,
    this.manufacturerData = const [],
    this.serviceData = const [],
    this.deviceName,
  });

  /// [fromPacket] creates a [List<BleAdvertisement>] from a raw message following this structure:
  /// MAC_ADDRESS;UNIX;LAT;LNG;ALT;MODEL;RSSI;MANUFACTURER+DATA;SERVICE+DATA;CRC16
  static List<BleAdvertisement> fromPacket(String raw) {
    final parts = raw.split(';');
    if (parts.isEmpty) return [];

    if (parts.length == 1) return [];
    // The parts should be divisible by 4, can be multiple groups

    if (parts.length % 12 != 0) {
      throw ParseException('Invalid advertisement definition');
    }

    // Separate each group
    final List<BleAdvertisement> advertisements = [];

    for (int i = 0; i < parts.length; i += 12) {
      final rawMacAddress = parts[i];
      final rawUnix = parts[i + 1];
      final rawLatitude = parts[i + 2];
      final rawLongitude = parts[i + 3];
      final rawAltitude = parts[i + 4];
      final model = parts[i + 5];
      final deviceName = parts[i + 6];
      final rawRssi = parts[i + 7];
      final rawTxPower = parts[i + 8];
      final rawManufacturerData = parts[i + 9];
      final rawServiceData = parts[i + 10];

      final receivedCrc = int.tryParse(parts[i + 11], radix: 16) ?? 0;

      final calculatedCrc = calculateCrc('${parts.sublist(i, i + 11).join(';')};'.codeUnits);
      if (receivedCrc != calculatedCrc) {
        throw CrcException(
          'Invalid CRC, expected ${receivedCrc.toRadixString(16)}, '
          'got ${calculatedCrc.toRadixString(16)}',
        );
      }

      final macParts = rawMacAddress.split('');
      if (macParts.length != 12) {
        throw MalformedException('Invalid MAC Address');
      }

      String macAddress = '';
      for (int i = 0; i < macParts.length; i += 2) {
        macAddress += '${macParts[i]}${macParts[i + 1]}';
        if (i + 2 < macParts.length) {
          macAddress += ':';
        }
      }

      DateTime timestamp;
      try {
        timestamp = DateTime.fromMillisecondsSinceEpoch(int.parse(rawUnix) * 1000);
      } catch (e) {
        throw MalformedException('Invalid timestamp');
      }

      double? latitude;
      if (rawLatitude.isNotEmpty) {
        latitude = double.tryParse(rawLatitude);
      }

      double? longitude;
      if (rawLongitude.isNotEmpty) {
        longitude = double.tryParse(rawLongitude);
      }

      double? altitude;
      if (rawAltitude.isNotEmpty) {
        altitude = double.tryParse(rawAltitude);
      }

      int rssi = int.tryParse(rawRssi) ?? 0;
      int? txPower = int.tryParse(rawTxPower);

      List<BleManufacturerData> manufacturerData = [];
      for (String mfd in rawManufacturerData.split(',')) {
        if (mfd.isEmpty) continue;
        final subparts = mfd.split(':');
        if (subparts.isEmpty) continue;

        if (subparts.length != 2) {
          throw MalformedException('Invalid manufacturer data "$mfd"');
        }

        final companyId = int.tryParse(subparts[0], radix: 16) ?? 0;
        List<int> data = [];
        for (int i = 0; i < subparts[1].length; i += 2) {
          data.add(int.tryParse(subparts[1].substring(i, i + 2), radix: 16) ?? 0);
        }
        manufacturerData.add(BleManufacturerData(companyId: companyId, data: data));
      }

      List<BleServiceData> serviceData = [];

      for (String sfd in rawServiceData.split(',')) {
        if (sfd.isEmpty) continue;
        final subparts = sfd.split(':');
        if (subparts.isEmpty) continue;

        if (subparts.length != 2) {
          throw MalformedException('Invalid service data "$sfd"');
        }

        final uuid = int.tryParse(subparts[0], radix: 16) ?? 0;
        List<int> data = [];
        for (int i = 0; i < subparts[1].length; i += 2) {
          data.add(int.tryParse(subparts[1].substring(i, i + 2), radix: 16) ?? 0);
        }

        serviceData.add(BleServiceData(uuid: uuid, data: data));
      }

      advertisements.add(
        BleAdvertisement(
          deviceName: deviceName,
          macAddress: macAddress,
          timestamp: timestamp,
          rssi: rssi,
          model: model,
          latitude: latitude,
          longitude: longitude,
          altitude: altitude,
          txPower: txPower,
          manufacturerData: manufacturerData,
          serviceData: serviceData,
        ),
      );
    }

    return advertisements;
  }

  /// [toPacket] converts the [BleAdvertisement] to a raw message following this structure:
  /// MAC_ADDRESS;UNIX;LAT;LNG;ALT;MODEL;RSSI;MANUFACTURER+DATA;SERVICE+DATA;CRC16
  String toPacket() {
    String payload = '';

    payload += '${macAddress.replaceAll(':', '').toUpperCase()};';
    payload += '${timestamp.millisecondsSinceEpoch ~/ 1000};';
    if (latitude != null) {
      payload += '$latitude;';
    } else {
      payload += ';';
    }
    if (longitude != null) {
      payload += '$longitude;';
    } else {
      payload += ';';
    }
    if (altitude != null) {
      payload += '$altitude;';
    } else {
      payload += ';';
    }
    payload += '$model;';
    if (deviceName != null) {
      payload += '$deviceName;';
    } else {
      payload += ';';
    }
    payload += '$rssi;';
    if (txPower != null) {
      payload += '$txPower;';
    } else {
      payload += ';';
    }
    payload += '${manufacturerData.map((e) => e.toPacket()).join(',')};';
    payload += '${serviceData.map((e) => e.toPacket()).join(',')};';

    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();

    return '$payload$crc';
  }
}

extension LayrzProtocolManufacturerDataSpec on BleManufacturerData {
  /// [toPacket] converts the [BleManufacturerData] to a raw message following this structure:
  /// MANUFACTURER_ID:DATA
  String toPacket() {
    String message = '${companyId.toRadixString(16).padLeft(4, '0').toUpperCase()}:';
    message += (data ?? []).map((e) => e.toRadixString(16).padLeft(2, '0')).join('').toUpperCase();
    return message;
  }
}

extension LayrzProtocolServiceDataSpec on BleServiceData {
  /// [toPacket] converts the [BleServiceData] to a raw message following this structure:
  /// SERVICE_UUID:DATA
  String toPacket() {
    String message = '${uuid.toRadixString(16).padLeft(4, '0').toUpperCase()}:';
    message += (data ?? []).map((e) => e.toRadixString(16).padLeft(2, '0')).join('').toUpperCase();
    return message;
  }
}
